java/clients/src/test/java/sleeper/clients/QueryWebSocketClientTest.java

```diff
@@ -0,0 +1,511 @@
+        return Query.builder()
+                .tableName(tableProperties.get(TABLE_NAME))
+                .queryId(queryId)
+                .regions(List.of(new Region(new Range(rowKey, min, true, max, false))))
+                .build();
+    }
+
+    private static String queryResult(String queryId, Record... records) {
+        return "{" +
+                "\"queryId\":\"" + queryId + "\", " +
+                "\"message\":\"records\"," +
+                "\"records\":[" + Stream.of(records).map(record -> "\"" + record + "\"").collect(Collectors.joining(",")) + "]" +
+                "}";
+    }
+
+    private static String completedQuery(String queryId, long recordCount) {
+        return "{" +
+                "\"queryId\":\"" + queryId + "\", " +
+                "\"message\":\"completed\"," +
+                "\"recordCount\":\"" + recordCount + "\"," +
+                "\"locations\":[{\"type\":\"websocket-endpoint\"}]" +
+                "}";
+    }
+
+    private static String createdSubQueries(String queryId, String... subQueryIds) {
+        return "{" +
+                "\"queryId\":\"" + queryId + "\", " +
+                "\"message\":\"subqueries\"," +
+                "\"queryIds\":[" + Stream.of(subQueryIds).map(id -> "\"" + id + "\"").collect(Collectors.joining(",")) + "]" +
+                "}";
+    }
+
+    private static String errorMessage(String queryId, String message) {
+        return "{" +
+                "\"queryId\":\"" + queryId + "\", " +
+                "\"message\":\"error\"," +
+                "\"error\":\"" + message + "\"" +
+                "}";
+    }
+
+    private static String unknownMessage(String queryId) {
+        return "{" +
+                "\"queryId\":\"" + queryId + "\"," +
+                "\"message\":\"unknown\"" +
+                "}";
+    }
+
+    protected void runQueryClient(String queryId, Client webSocketClient) throws Exception {
+        new QueryWebSocketClient(instanceProperties, tableIndex, new FixedTablePropertiesProvider(tableProperties),
+                in.consoleIn(), out.consoleOut(), webSocketClient, () -> queryId)
+                .run();
+    }
+
+    private FakeWebSocketClient withResponses(WebSocketResponse... responses) {
+        client = new FakeWebSocketClient(new FixedTablePropertiesProvider(tableProperties), out.consoleOut());
+        client.withResponses(responses);
+        return client;
+    }
+
+    public WebSocketResponse open(Query query) {
+        return messageHandler -> messageHandler.onOpen(query, client::send);
+    }
+
+    private WebSocketResponse message(String message) {
+        return messageHandler -> messageHandler.onMessage(message);
+    }
+
+    public WebSocketResponse closeWithReason(String reason) {
+        return messageHandler -> messageHandler.onClose(reason);
+    }
+
+    public WebSocketResponse error(Exception error) {
+        return messageHandler -> messageHandler.onError(error);
+    }
```

Comment:
This method is unused. Should there be a test for this?

Note that onClose will be called after onError.

Reply:
Sorry, it was a failure in my IDE. We could do with testing when onError is called with or without onClose.